docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+### Why are top-level modules always unbound?
+
+When we call `apply`, a copy of the top-level Module is created which will actually hold the variables and PRNG sequences. This stateful, "bound", clone only exists while we are executing the apply method. The reason for this is that if you create a stateful object and destroy it before the apply function returns, the `apply` function itself behaves like a pure function (pure means if you put the same arguments in, it will return the same outputs).
+
+Pure functions have many advantages but when using JAX they are often essential. For example, most code requires compilation using `jax.jit` to be fast and once you created a Module you probably want to optimize its parameters using `jax.grad`. However, these APIs expect a pure function and don't work on stateful bound `Module` instances directly. Moreover, pure functions allow for flexible interoperability with other libraries.
+
+### Cloning
+
+To make this approach work reliably we need well-defined cloning behavior. Rather than relying on a complex nested cloning procedure like Python's `deepcopy`, Flax enforces that a `Module` is exactly defined by its construction arguments. Therefore cloning a Module reduces to calling the constructor with its original construction arguments. Because `Module` acts as an immutable dataclass, the construction arguments are mapped 1:1 to instance attributes.
+
+### Bind
+
+Sometimes it's useful to have a bound, top-level Module without having to wrap the code in a function. For example: to interact with a Module inside a Jupyter notebook. The [bind](https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=bind#flax.linen.Module.bind) method returns a bound clone with an unlimited lifetime. The downside of this is that you cannot combine it with JAX transformations or integrate it into a vanilla JAX codebase that expects stateless code.
+
+
+## Setup
+
+The `setup` method is often used like the constructor hook (`__init__`) in normal Python classes. However, for more advanced use cases it’s good to realize that it is not quite the same as a constructor.
+
+`setup` is only called after a Module becomes bound. Normally, this is not an issue because most Modules are bound (almost) immediately (as part of `init` and `apply`). Inside `setup`, sub-modules become bound when they are assigned to an attribute. nside an `nn.compact` decorated method, sub-modules are bound immediately when constructed. As explained in the previous section, top-level Modules are never bound and thus setup is not called when they are constructed. This means that you cannot access attributes assigned in setup from an unbound, top-level module.
```

Comment:
```suggestion
`setup` is only called after a Module becomes bound. Normally, this is not an issue because most Modules are bound (almost) immediately (as part of `init` and `apply`). Inside `setup`, sub-modules become bound when they are assigned to an attribute. Inside an `nn.compact` decorated method, sub-modules are bound immediately when constructed. As explained in the previous section, top-level Modules are never bound and thus setup is not called when they are constructed. This means that you cannot access attributes assigned in setup from an unbound, top-level module.
```